File: solutions/r/robot-simulator/1/robot-simulator.R
new_robot <- function(coordinates, direction) {
  structure(
    list(
      coordinates = coordinates,
      direction = direction
    ),
    class = "robot"
  )
}

move <- function(a_robot, commands) {
  UseMethod("move")
}

# nolint start
move.robot <- function(a_robot, commands) {
  directions <- c("NORTH", "EAST", "SOUTH", "WEST")

  turn_right <- function(direction) {
    current_index <- match(direction, directions)
    directions[(current_index %% length(directions)) + 1]
  }

  turn_left <- function(direction) {
    current_index <- match(direction, directions)
    directions[((current_index - 2) %% length(directions)) + 1]
  }

  advance <- function(robot) {
    if (robot$direction == "NORTH") {
      robot$coordinates[2] <- robot$coordinates[2] + 1
    } else if (robot$direction == "SOUTH") {
      robot$coordinates[2] <- robot$coordinates[2] - 1
    } else if (robot$direction == "EAST") {
      robot$coordinates[1] <- robot$coordinates[1] + 1
    } else if (robot$direction == "WEST") {
      robot$coordinates[1] <- robot$coordinates[1] - 1
    }

    robot
  }

  instructions <- strsplit(commands, "")[[1]]

  for (instruction in instructions) {
    if (instruction == "R") {
      a_robot$direction <- turn_right(a_robot$direction)
    } else if (instruction == "L") {
      a_robot$direction <- turn_left(a_robot$direction)
    } else if (instruction == "A") {
      a_robot <- advance(a_robot)
    }
  }

  a_robot
}
# nolint end

File: solutions/cpp/robot-simulator/1/robot_simulator.cpp
#include "robot_simulator.h"
namespace robot_simulator {
  auto Robot::get_position() const -> pair<int, int> {
    return make_pair(x, y);
  }
  auto Robot::get_bearing() const -> Bearing {
    return bearing;
  }
  auto Robot::turn_right() -> void {
    switch(bearing) {
      case Bearing::NORTH:
        bearing = Bearing::EAST;
        break;
      case Bearing::EAST:
        bearing = Bearing::SOUTH;
        break;
      case Bearing::SOUTH:
        bearing = Bearing::WEST;
        break;
      default:
        bearing = Bearing::NORTH;
        break;
    }
  }
  auto Robot::turn_left() -> void {
    switch(bearing) {
      case Bearing::NORTH:
        bearing = Bearing::WEST;
        break;
      case Bearing::WEST:
        bearing = Bearing::SOUTH;
        break;
      case Bearing::SOUTH:
        bearing = Bearing::EAST;
        break;
      default:
        bearing = Bearing::NORTH;
        break;
    }
  }
  auto Robot::advance() -> void {
    switch(bearing) {
      case Bearing::NORTH:
        y += 1;
        break;
      case Bearing::EAST:
        x += 1;
        break;
      case Bearing::SOUTH:
        y -= 1;
        break;
      default:
        x -= 1;
        break;
    }
  }
  auto Robot::execute_sequence(const string& sequence) -> void {
    for(char step : sequence) {
      switch(step) {
        case 'R':
          turn_right();
          break;
        case 'L':
          turn_left();
          break;
        case 'A':
          advance();
          break;
      }
    }
  }
}
